keyboards/dztech/dz65rgb/keymaps/drootz/keymap.c

```diff
@@ -0,0 +1,774 @@
+    } 
+
+    /* Current layer LED indicator */
+    rgb_matrix_set_color(layers_leds_map[biton32(layer_state)], WHITE_HSV_DIM[0], WHITE_HSV_DIM[1], WHITE_HSV_DIM[2]); 
+
+    /* Leader Key LED under-glow */
+    if (isLeader) {
+        rgb_matrix_set_color(14, CONFIG_HSV[0], CONFIG_HSV[1], CONFIG_HSV[2]);
+        rgb_matrix_set_color(30, CONFIG_HSV[0], CONFIG_HSV[1], CONFIG_HSV[2]);
+    } else {
+        rgb_matrix_set_color(14, CONFIG_HSV_DIM[0], CONFIG_HSV_DIM[1], CONFIG_HSV_DIM[2]);
+    }
+
+    /* Blinking LED indicator when recording Dynamic Macro */
+    if (isRecording && isBlinking) {
+        get_this_led_blinking(0, false, 0, 255);
+    }
+
+    /* Blinking LED indicator when playing Dynamic Macro */
+    if (isPlaying && isBlinking) {
+        if (timer_elapsed(macro_one_play_timer) < macro_play_blink_timer || timer_elapsed(macro_two_play_timer) < macro_play_blink_timer) {
+            get_this_led_blinking(0, true, 85, 255);
+        } else {
+            if (isBlinking && isPlaying) {
+                reset_blink_status();
+                reset_blink_cycle();
+            }
+        }
+    }
+
+    switch (biton32(layer_state)) {
+        case _FN:
+            /* Dynamic Macro LED indicator */
+            if (isRecording) {
+                rgb_matrix_set_color(0, RED_HSV[0], RED_HSV[1], RED_HSV[2]); /* macro stop */
+            } else {
+                rgb_matrix_set_color(43, RED_HSV_DIM[0], RED_HSV_DIM[1], RED_HSV_DIM[2]); /* macro 1 record */
+                rgb_matrix_set_color(29, RED_HSV_DIM[0], RED_HSV_DIM[1], RED_HSV_DIM[2]); /* macro 2 record */
+                rgb_matrix_set_color(42, GREEN_HSV_DIM[0], GREEN_HSV_DIM[1], GREEN_HSV_DIM[2]); /* macro 1 play */
+                rgb_matrix_set_color(28, GREEN_HSV_DIM[0], GREEN_HSV_DIM[1], GREEN_HSV_DIM[2]); /* macro 2 play */
+            }
+            /* Layer LED indicators */
+            rgb_matrix_set_color(45, WHITE_HSV_DIM[0], WHITE_HSV_DIM[1], WHITE_HSV_DIM[2]); /* Layer _MAIN */
+            rgb_matrix_set_color(46, WHITE_HSV_DIM[0], WHITE_HSV_DIM[1], WHITE_HSV_DIM[2]); /* Layer _MAC */
+            break; 
+    }
+
+    #endif /* RGB_MATRIX */
+}
+ 
+bool process_record_user(uint16_t keycode, keyrecord_t *record) {
+    bool pressed = record->event.pressed;
+    if (pressed) {
+        switch (keycode) {
+            case TO(_MAC):
+                onMac = true;
+                break;
+            case TO(_MAIN):
+                onMac = false;      
+                break;
+        }
+    }
+    return true;
+} 
+
+//**************** LEADER *********************//
+
+#ifdef LEADER_ENABLE
+
+LEADER_EXTERNS();
+
+/* éÉ */
+void send_e_aigu(void) {
+    if (IS_HOST_LED_ON(USB_LED_CAPS_LOCK)) {
+        if (onMac) {
+            SEND_STRING(SS_TAP(X_CAPSLOCK) SS_LALT("e") SS_LSFT("e"));
+            registerCapsLock();
+        } else {
+            SEND_STRING(SS_DOWN(X_LALT) SS_TAP(X_P0) SS_TAP(X_P2) SS_TAP(X_P0) SS_TAP(X_P1) SS_UP(X_LALT));
```

Comment:
this would be simpler .... 

https://docs.qmk.fm/#/feature_unicode

Reply:
Right, I thought of using this feature. Here's why I decided not to implement it.

I use multiple computers on a regualr basis and some are locked down corporate properties(Win). This means that I am unable to install any softwares or change registries. The problem I am trying to solve is to be able to output a subset of French accents using the same Leader Key sequences on a standard English keyboard input setting on both Win and Mac OS. (without installing softwares or tweaking input settings)

Reasons for not implementing the QMK unicode feature:

1. QMK unicode requires OS specific configurations (ex.: Win: registry change, Win: WinCompose, Mac: Unicode Input Settings)
	- I need something to work "out of the box" without any additional OS specific configurations and need to be able to use same Leader Key sequences on both Win and Mac to output the same French accent.
2. When I initially tested the unicode feature, I was not able to trigger unicodes keycodes from Leader Key macros. It only worked from keymap layers which is not the workflow I am looking for. _(Surely I am the problem here but if there is a way to trigger unicodes from macro I did not find it)_
3. Firmware size: I have quite a lot of Leader Key macro.(70+) Enabling the unicode feature increase the firmware size just enough to be a problem, even if let's say, I use the feature partially for Mac outputs only.

In other words, I am trying to implement a work around the QMK unicodes feature and bind a subset of French accents to Leader Key sequences that would have the same output on both Win and Mac computers "out of the box" and without any OS configurations.

Example:
```
1. LeaderKey > E  // output è/È
2. LeaderKey > E > E  // output ê/Ê
3. LeaderKey > E > E > E  // output é/É
```
Hope this make sense...?

Reply:
I’ll simplify this block in my next commit by removing repetition.

Reply:
You can still use the unicode feature to simplify things.   The "start" and "stop" functions, (as well as the "hex to keycode") functions are defined weakly, meaning that you can replace them.  And replace them with custom code.  This would streamline things, as you could use `UC(0x0201)`, basically. 

https://github.com/qmk/qmk_firmware/blob/42d6270f28831e95d1cb9c14a7423d5b1d864d67/quantum/process_keycode/process_unicode_common.c#L76-L134

Reply:
I understand that I can change/control what happens before and after the unicode register with the weakly typed functions "start"/"stop". Not sure how I can change/control the actual "code registration" though. Obviously I still can't wrap my head around how to leverage unicodes without some kind of configuration on the host computer. I tried couple of things with no success. Until then, I will use "decimal" unicodes instead of hex. (ex.: Alt + KC_P1...3...1). This way, I do not need special software or setting on the host computer.

I did manage to simplify and remove repetition for the "french accents" code block in commit 971fb94. I am not a professional programmer so I hope this part will be ok for now. The firmware compile fine and works. Was even able to shave off a couple of bytes in the process.

btw thanks for the feedback, really appreciated.

Reply:
Well, for the decimal stuff, I think if you replace this function: 
https://github.com/qmk/qmk_firmware/blob/42d6270f28831e95d1cb9c14a7423d5b1d864d67/quantum/process_keycode/process_unicode_common.c#L136-L144

That's may be what you're looking for.  Since this controls what characters are used.

Eg, if you add this to your keymap:
```c
uint16_t hex_to_keycode(uint8_t hex) {
    if (hex == 0x0) {
        return KC_P0;
    } else if (hex < 0xA) {
        return KC_P1 + (hex - 0x1);
    } else {
        return KC_A + (hex - 0xA);
    }
}
```

That should use the numpad keys.  Though, one caveat here, you may want/need to ensure that numlock is enabled to ensure the correct keys get pressed.

Reply:
Looks like unicode is not yet part of macro commands:

https://github.com/qmk/qmk_firmware/blob/85041ff05bf0e5f4ff4535caf6e638491a5614c8/tmk_core/common/action_macro.h#L70-L79

This may be why I am unable to make it work with Leader Key macros. I can't find an implementation for it.

Reply:
That's the TMK macros, which are deprecated.